File: algorithms/algorithms-cplusplus/leetcode/Question_1052_Grumpy_Bookstore_Owner.cc
#include <vector>

using namespace std;

class Solution {
 public:
  int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int minutes) {
    int satisfied = 0, m_add_satisfied = 0, add_satisfied = 0;
    for (int i = 0; i < customers.size(); ++i) {
      satisfied += !grumpy[i] * customers[i];
      add_satisfied += grumpy[i] * customers[i];
      if (i >= minutes) {
        add_satisfied -= grumpy[i - minutes] * customers[i - minutes];
      }
      m_add_satisfied = max(m_add_satisfied, add_satisfied);
    }
    return satisfied + m_add_satisfied;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_1052_Grumpy_Bookstore_Owner.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_1052_Grumpy_Bookstore_Owner {
    public int maxSatisfied(int[] customers, int[] grumpy, int minutes) {
        int satisfied = 0, maxMakeSatisfied = 0;
        for (int i = 0, winOfMakeSatisfied = 0; i < grumpy.length; ++i) {
            if (grumpy[i] == 0) {
                satisfied += customers[i];
            } else {
                winOfMakeSatisfied += customers[i];
            }
            if (i >= minutes) {
                winOfMakeSatisfied -= grumpy[i - minutes] * customers[i - minutes];
            }
            maxMakeSatisfied = Math.max(winOfMakeSatisfied, maxMakeSatisfied);
        }
        return satisfied + maxMakeSatisfied;
    }
}
